test/upgrade/types.go

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2020 The Knative Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package upgrade
+
+import (
+	"testing"
+
+	"go.uber.org/zap"
+)
+
+// Suite represents a upgrade tests suite that can be executed and will perform
+// execution in predictable manner
+type Suite struct {
+	Tests         Tests
+	Installations Installations
+}
+
+// Tests holds a list of operations for various part of upgrade suite
+type Tests struct {
+	PreUpgrade     []Operation
+	PostUpgrade    []Operation
+	PostDowngrade  []Operation
+	ContinualTests []BackgroundOperation
+}
+
+// Installations holds a list of operations that will install Knative components
+// in different versions
+type Installations struct {
+	Base          []Operation
+	UpgradeWith   []Operation
+	DowngradeWith []Operation
+}
+
+// Named interface represents a type that is named in user friendly way
+type Named interface {
+	Name() string
+}
+
+// Operation represents a upgrade test operation like test or installation that
+// can be provided by specific component or reused in aggregating components
+type Operation interface {
+	Named
+	Handler() func(c Context)
+}
+
+// BackgroundOperation represents a upgrade test operation that will be
+// performed in background while other operations is running. To achieve that
+// a passed BackgroundContext should be used to synchronize it's operations with
+// Ready and Stop channels.
+type BackgroundOperation interface {
+	Named
```

Comment:
Could we just just `Name() ` here rather than importing a type that has one method Name(). Seems like this contruct is actually never leveraged and it's a bit confusing why it is done this way. Same for the `Operation` type.